usr/src/man/man1/touch.1
.\"
.\" Sun Microsystems, Inc. gratefully acknowledges The Open Group for
.\" permission to reproduce portions of its copyrighted documentation.
.\" Original documentation from The Open Group can be obtained online at
.\" http://www.opengroup.org/bookstore/.
.\"
.\" The Institute of Electrical and Electronics Engineers and The Open
.\" Group, have given us permission to reprint portions of their
.\" documentation.
.\"
.\" In the following statement, the phrase ``this text'' refers to portions
.\" of the system documentation.
.\"
.\" Portions of this text are reprinted and reproduced in electronic form
.\" in the SunOS Reference Manual, from IEEE Std 1003.1, 2004 Edition,
.\" Standard for Information Technology -- Portable Operating System
.\" Interface (POSIX), The Open Group Base Specifications Issue 6,
.\" Copyright (C) 2001-2004 by the Institute of Electrical and Electronics
.\" Engineers, Inc and The Open Group.  In the event of any discrepancy
.\" between these versions and the original IEEE and The Open Group
.\" Standard, the original IEEE and The Open Group Standard is the referee
.\" document.  The original Standard can be obtained online at
.\" http://www.opengroup.org/unix/online.html.
.\"
.\" This notice shall appear on any product containing this material.
.\"
.\" The contents of this file are subject to the terms of the
.\" Common Development and Distribution License (the "License").
.\" You may not use this file except in compliance with the License.
.\"
.\" You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
.\" or http://www.opensolaris.org/os/licensing.
.\" See the License for the specific language governing permissions
.\" and limitations under the License.
.\"
.\" When distributing Covered Code, include this CDDL HEADER in each
.\" file and include the License file at usr/src/OPENSOLARIS.LICENSE.
.\" If applicable, add the following below this CDDL HEADER, with the
.\" fields enclosed by brackets "[]" replaced with your own identifying
.\" information: Portions Copyright [yyyy] [name of copyright owner]
.\"
.\"
.\" Copyright 1989 AT&T
.\" Copyright (c) 1992, X/Open Company Limited All Rights Reserved
.\" Portions Copyright (c) 2009, Sun Microsystems, Inc. All Rights Reserved
.\"
.TH TOUCH 1 "June 13, 2021"
.SH NAME
touch, settime \- change file access and modification times
.SH SYNOPSIS
.nf
\fBtouch\fR [\fB-acm\fR] [\fB-r\fR \fIref_file\fR | \fB-t\fR \fItime\fR | \fB-d\fR \fIdate_time\fR] \fIfile\fR...
.fi

.LP
.nf
\fBtouch\fR [\fB-acm\fR] [\fItime_spec\fR] \fIfile\fR...
.fi

.LP
.nf
\fBsettime\fR [\fB-f\fR \fIref_file\fR] [\fItime_spec\fR] \fIfile\fR...
.fi

.SH DESCRIPTION
The \fBtouch\fR utility sets the access and modification times of each file.
The \fIfile\fR operand is created if it does not already exist.
.sp
.LP
The time used can be specified by \fB-t\fR \fItime\fR, by \fB-d\fR
\fIdate_time\fR, by the corresponding time fields of the file referenced by
\fB-r\fR \fIref_file\fR, or by the \fItime_spec\fR operand. If none of these
are specified, \fBtouch\fR uses the current time.
.sp
.LP
If neither the \fB-a\fR nor \fB-m\fR options are specified, \fBtouch\fR updates
both the modification and access times.
.sp
.LP
A user with write access to a file, but who is not the owner of the file or a
super-user, can change the modification and access times of that file only to
the current time. Attempts to set a specific time with \fBtouch\fR results in
an error.
.sp
.LP
The \fBsettime\fR utility is equivalent to \fBtouch\fR \fB-c\fR
[\fItime_spec\fR] \fIfile\fR.
.SH OPTIONS
The following options are supported in the \fBtouch\fR and \fBsettime\fR
utilities:
.SS "touch"
The following options are supported for the \fBtouch\fR utility:
.sp
.ne 2
.na
\fB\fB-a\fR\fR
.ad
.sp .6
.RS 4n
Changes the access time of \fIfile\fR. Does not change the modification time
unless \fB-m\fR is also specified.
.RE

.sp
.ne 2
.na
\fB\fB-c\fR\fR
.ad
.sp .6
.RS 4n
Does not create a specified \fIfile\fR if it does not exist. Does not write any
diagnostic messages concerning this condition.
.RE

.sp
.ne 2
.na
\fB\fB-d\fR \fIdate_time\fR\fR
.ad
.sp .6
.RS 4n
Uses the specified date_time instead of the current time. The option-argument
must be a string of the form:
.sp
.in +2
.nf
\fIYYYY\fR-\fIMM\fR-\fIDDThh\fR:\fImm\fR:\fISS\fR[.\fIfrac\fR][\fItz\fR]
.fi
.in -2
.sp

or
.sp
.in +2
.nf
\fIYYYY\fR-\fIMM\fR-\fIDDThh\fR:\fImm\fR:\fISS\fR[,\fIfrac\fR][\fItz\fR]
.fi
.in -2
.sp

where
.RS +4
.TP
.ie t \(bu
.el o
\fIYYYY\fR is at least four decimal digits giving the year
.RE
.RS +4
.TP
.ie t \(bu
.el o
\fIMM\fR, \fIDD\fR, \fIhh\fR, \fImm\fR, and \fISS\fR are as with \fB-t\fR
\fItime\fR
.RE
.RS +4
.TP
.ie t \(bu
.el o
\fIT\fR is either the letter \fBT\fR or a single SPACE character
.RE
.RS +4
.TP
.ie t \(bu
.el o
[\fI\&.frac\fR] and [\fI,frac\fR] are either empty, or a period (\fB\&.\fR) or
a comma (\fB,\fR) respectively, followed by one or more decimal digits,
specifying a fractional second
.RE
.RS +4
.TP
.ie t \(bu
.el o
[\fItz\fR] is either empty, signifying local time, or the letter \fBZ\fR,
signifying UTC. If [\fItz\fR] is empty, the resulting time is affected by the
value of the \fBTZ\fR environment variable
.RE
.RE

.sp
.ne 2
.na
\fB\fB-m\fR\fR
.ad
.sp .6
.RS 4n
Changes the modification time of \fIfile\fR. Does not change the access time
unless \fB-a\fR is also specified.
.RE

.sp
.ne 2
.na
\fB\fB-r\fR \fIref_file\fR\fR
.ad
.sp .6
.RS 4n
Uses the corresponding times of the file named by \fIref_file\fR instead of the
current time.
.RE

.sp
.ne 2
.na
\fB\fB-t\fR \fItime\fR\fR
.ad
.sp .6
.RS 4n
Uses the specified \fItime\fR instead of the current time. \fItime\fR is a
decimal number of the form:
.sp
.in +2
.nf
[[\fICC\fR]\fIYY\fR]\fIMMDDhhmm\fR[\fI\&.SS\fR]
.fi
.in -2
.sp

where each two digits represent the following:
.sp
.ne 2
.na
\fB\fIMM\fR\fR
.ad
.sp .6
.RS 4n
The month of the year [01-12].
.RE

.sp
.ne 2
.na
\fB\fIDD\fR\fR
.ad
.sp .6
.RS 4n
The day of the month [01-31].
.RE

.sp
.ne 2
.na
\fB\fIhh\fR\fR
.ad
.sp .6
.RS 4n
The hour of the day [00-23].
.RE

.sp
.ne 2
.na
\fB\fImm\fR\fR
.ad
.sp .6
.RS 4n
The minute of the hour [00-59].
.RE

.sp
.ne 2
.na
\fB\fICC\fR\fR
.ad
.sp .6
.RS 4n
The first two digits of the year.
.RE

.sp
.ne 2
.na
\fB\fIYY\fR\fR
.ad
.sp .6
.RS 4n
The second two digits of the year.
.RE

.sp
.ne 2
.na
\fB\fISS\fR\fR
.ad
.sp .6
.RS 4n
The second of the minute [00-61].
.RE

Both \fICC\fR and \fIYY\fR are optional. If neither is given, the current year
is assumed. If \fIYY\fR is specified, but \fICC\fR is not, \fICC\fR is derived
as follows:
.sp

.sp
.TS
box;
c c
l l .
If YY is:	CC becomes:
_
69-99	19
00-38	20
39-68	ERROR
.TE

The resulting time is affected by the value of the \fBTZ\fR environment
variable. If the resulting time value precedes the Epoch, \fBtouch\fR exits
immediately with an error status. The range of valid times is the Epoch to
January 18, 2038.
.sp
The range for \fISS\fR is [00-61] rather than [00-59] because of leap seconds.
If \fISS\fR is 60 or 61, and the resulting time, as affected by the \fBTZ\fR
environment variable, does not refer to a leap second, the resulting time is
one or two seconds after a time where \fISS\fR is 59. If \fISS\fR is not given,
it is assumed to be 0.
.RE

.SS "settime"
The following option is supported for the \fBsettime\fR utility:
.sp
.ne 2
.na
\fB\fB-f\fR \fIref_file\fR\fR
.ad
.sp .6
.RS 4n
Uses the corresponding times of the file named by \fIref_file\fR instead of the
current time.
.RE

.SH OPERANDS
The following operands are supported for the \fBtouch\fR and \fBsettime\fR
utilities:
.sp
.ne 2
.na
\fB\fIfile\fR\fR
.ad
.sp .6
.RS 4n
A path name of a file whose times are to be modified.
.RE

.sp
.ne 2
.na
\fB\fItime_spec\fR\fR
.ad
.sp .6
.RS 4n
Uses the specified \fItime_spec\fR instead of the current time. This operand is
a decimal number of the form:
.sp
.in +2
.nf
\fIMMDDhhmm\fR[\fIYY\fR]
.fi
.in -2
.sp

where each two digits represent the following:
.sp
.ne 2
.na
\fB\fIMM\fR\fR
.ad
.sp .6
.RS 4n
The month of the year [01-12].
.RE

.sp
.ne 2
.na
\fB\fIDD\fR\fR
.ad
.sp .6
.RS 4n
The day of the month [01-31].
.RE

.sp
.ne 2
.na
\fB\fIhh\fR\fR
.ad
.sp .6
.RS 4n
The hour of the day [00-23].
.RE

.sp
.ne 2
.na
\fB\fImm\fR\fR
.ad
.sp .6
.RS 4n
The minute of the hour [00-59].
.RE

.sp
.ne 2
.na
\fB\fIYY\fR\fR
.ad
.sp .6
.RS 4n
The second two digits of the year.
.sp
\fIYY\fR is optional. If it is omitted, the current year is assumed. If
\fIYY\fR is specified, the year is derived as follows:
.RE

.sp

.sp
.TS
box;
c c
l l .
YY	Corresponding Year
69-99	1969-1999
00-38	2000-2038
39-68	ERROR
.TE

If no \fB-d\fR, \fB-r\fR, or \fB-t\fR option is specified, at least two operands
are specified, and the first operand is an eight- or ten-digit decimal integer,
the first operand is assumed to be a \fItime_spec\fR operand. Otherwise, the
first operand is assumed to be a \fIfile\fR operand.
.RE

.SH USAGE
See \fBlargefile\fR(7) for the description of the behavior of \fBtouch\fR when
encountering files greater than or equal to 2 Gbyte ( 2^31 bytes).
.SH ENVIRONMENT VARIABLES
See \fBenviron\fR(7) for descriptions of the following environment variables
that affect the execution of \fBtouch\fR: \fBLANG\fR, \fBLC_ALL\fR, \fBLANG\fR,
\fBLC_ALL\fR, \fBLC_CTYPE\fR, \fBLC_MESSAGES\fR, and \fBNLSPATH\fR.
.sp
.ne 2
.na
\fB\fBTZ\fR\fR
.ad
.sp .6
.RS 4n
Determine the timezone to be used for interpreting the \fItime\fR or
\fIdate_time\fR option-argument or the \fItime_spec\fR operand.
.RE

.SH EXIT STATUS
The following exit values are returned:
.sp
.ne 2
.na
\fB\fB0\fR\fR
.ad
.sp .6
.RS 4n
The \fBtouch\fR utility executed successfully and all requested changes were
made.
.RE

.sp
.ne 2
.na
\fB\fB>0\fR\fR
.ad
.sp .6
.RS 4n
An error occurred. The \fBtouch\fR utility returned the number of files for
which the times could not be successfully modified.
.RE

.SH ATTRIBUTES
See \fBattributes\fR(7) for descriptions of the following attributes:
.sp

.sp
.TS
box;
c | c
l | l .
ATTRIBUTE TYPE	ATTRIBUTE VALUE
_
CSI	Enabled
_
Interface Stability	Committed
_
Standard	See \fBstandards\fR(7).
.TE

.SH SEE ALSO
.BR futimens (2),
.BR stat (2),
.BR attributes (7),
.BR environ (7),
.BR largefile (7),
.BR standards (7)
.SH NOTES
Users familiar with the BSD environment find that for the \fBtouch\fR utility,
the \fB-f\fR option is accepted but ignored. The \fB-f\fR option is unnecessary
because \fBtouch\fR succeeds for all files owned by the user regardless of the
permissions on the files.


usr/src/ucbcmd/touch/touch.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License, Version 1.0 only
 * (the "License").  You may not use this file except in compliance
 * with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*
 * Copyright 2002 Sun Microsystems, Inc.  All rights reserved.
 * Use is subject to license terms.
 */

#ifndef lint
#pragma ident	"%Z%%M%	%I%	%E% SMI"
#endif

#include <stdio.h>
#include <ctype.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/file.h>
#include <sys/time.h>
#include <time.h>
#include <errno.h>
#include <unistd.h>

#define	isleap(y) (((y) % 4) == 0 && ((y) % 100) != 0 || ((y) % 400) == 0)

struct	stat	stbuf;
int	status;
#ifdef S5EMUL
int dmsize[12]={31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
#endif /* S5EMUL */

static char usage[] =
#ifdef S5EMUL
		"[-amc] [mmddhhmm[yy]]";
#else /*!S5EMUL*/
		"[-amcf]";
int	force = 0;
int	nowrite;
#endif /*!S5EMUL*/

int	mflg=1, aflg=1, cflg=0, nflg=0;
char	*prog;

#ifdef S5EMUL
char	*cbp;
#endif /* S5EMUL */
time_t	time();
off_t	lseek();
time_t	timelocal(), timegm();
struct timeval timbuf;
static void timestruc_to_timeval(timestruc_t *, struct timeval *);

#ifdef S5EMUL
struct tm *
gtime()
{
	static struct tm newtime;
	long nt;

	newtime.tm_mon = gpair() - 1;
	newtime.tm_mday = gpair();
	newtime.tm_hour = gpair();
	if (newtime.tm_hour == 24) {
		newtime.tm_hour = 0;
		newtime.tm_mday++;
	}
	newtime.tm_min = gpair();
	newtime.tm_sec = 0;
	newtime.tm_year = gpair();
	if (newtime.tm_year < 0) {
		(void) time(&nt);
		newtime.tm_year = localtime(&nt)->tm_year;
	}
	return (&newtime);
}

gpair()
{
	int c, d;
	char *cp;

	cp = cbp;
	if (*cp == 0)
		return (-1);
	c = (*cp++ - '0') * 10;
	if (c<0 || c>100)
		return (-1);
	if (*cp == 0)
		return (-1);
	if ((d = *cp++ - '0') < 0 || d > 9)
		return (-1);
	cbp = cp;
	return (c+d);
}
#endif /*S5EMUL*/

int
main(int argc, char *argv[])
{
	int c;
#ifdef S5EMUL
	int days_in_month;
	struct tm *tp;
#endif /* S5EMUL */

	int errflg=0, optc;
	extern char *optarg;
	extern int optind;
	extern int opterr;

	prog = argv[0];
	opterr = 0;			/* disable getopt() error msgs */
	while ((optc=getopt(argc, argv, "amcf")) != EOF)
		switch (optc) {
		case 'm':
			mflg++;
			aflg--;
			break;
		case 'a':
			aflg++;
			mflg--;
			break;
		case 'c':
			cflg++;
			break;
#ifndef S5EMUL
		case 'f':
			force++;	/* SysV version ignores -f */
			break;
#endif /*!S5EMUL*/
		case '?':
			errflg++;
		}

	if (((argc-optind) < 1) || errflg) {
		(void) fprintf(stderr, "usage: %s %s file ...\n", prog, usage);
		exit(2);
	}
	status = 0;

#ifdef S5EMUL
	if (!isnumber(argv[optind])) {	/* BSD version only sets Present */
#endif /*S5EMUL*/
		if ((aflg <= 0) || (mflg <= 0))
			(void) gettimeofday(&timbuf, NULL);
		else
			nflg++;		/* no -a, -m, or date seen */
#ifdef S5EMUL
	} else {			/* SysV version sets arbitrary date */
		cbp = (char *)argv[optind++];
		if ((tp = gtime()) == NULL) {
			(void) fprintf(stderr, "%s: bad date conversion\n",
			    prog);
			exit(2);
		}
		days_in_month = dmsize[tp->tm_mon];
		if (tp->tm_mon == 1 && isleap(tp->tm_year + 1900))
			days_in_month = 29;	/* February in leap year */
		if (tp->tm_mon < 0 || tp->tm_mon > 11 ||
		    tp->tm_mday < 1 || tp->tm_mday > days_in_month ||
		    tp->tm_hour < 0 || tp->tm_hour > 23 ||
		    tp->tm_min < 0 || tp->tm_min > 59 ||
		    tp->tm_sec < 0 || tp->tm_sec > 59) {
			(void) fprintf(stderr, "%s: bad date conversion\n",
			    prog);
			exit(2);
		}
		timbuf = timelocal(tp);
	}
#endif /*S5EMUL*/

	for (c = optind; c < argc; c++) {
		if (touch(argv[c]) < 0)
			status++;
	}
	return (status);
}

int
touch(filename)
	char *filename;
{
	struct timeval times[2];
	int fd;

	if (stat(filename, &stbuf)) {
		/*
		 * if stat failed for reasons other than ENOENT,
		 * the file should not be created, since this
		 * can clobber the contents of an existing file
		 * (for example, a large file that results in overflow).
		 */
		if (errno != ENOENT) {
			(void) fprintf(stderr,"%s: cannot stat ", prog);
			perror(filename);
			return (-1);
		} else if (cflg) {
			return (-1);
		}
		else if ((fd = creat(filename, 0666)) < 0) {
			(void) fprintf(stderr, "%s: cannot create ", prog);
			perror(filename);
			return (-1);
		}
		else {
			(void) close(fd);
			if (stat(filename, &stbuf)) {
				(void) fprintf(stderr,"%s: cannot stat ", prog);
				perror(filename);
				return (-1);
			}
		}
		if (nflg)
			return (0);
	}

	times[0] = times[1] = timbuf;
	if (mflg <= 0)
		timestruc_to_timeval(&stbuf.st_mtim, times + 1);
	if (aflg <= 0)
		timestruc_to_timeval(&stbuf.st_atim, times);

#ifndef S5EMUL
	/*
	 * Since utime() allows the owner to change file times without
	 * regard to access permission, enforce BSD semantics here
	 * (cannot touch if read-only and not -f).
	 */
	nowrite = access(filename, R_OK|W_OK);
	if (nowrite && !force) {
		(void) fprintf(stderr,
		    "%s: cannot touch %s: no write permission\n",
		    prog, filename);
		return (-1);
	}
#endif /*!S5EMUL*/

	if (utimes(filename, nflg ? NULL : times)) {
		if (nflg && (errno != EROFS) && (errno != EACCES)) {
			/*
			 * If utime() failed to set the Present, it
			 * could be a BSD server that is complaining.
			 * If that's the case, try the old read/write trick.
			 */
			return (oldtouch(filename, &stbuf));
		}
		(void) fprintf(stderr,"%s: cannot change times on ", prog);
		perror(filename);
		return (-1);
	}
	return (0);
}

int
oldtouch(filename, statp)
	char *filename;
	struct stat *statp;
{
	int rwstatus;

	if ((statp->st_mode & S_IFMT) != S_IFREG) {
		(void) fprintf(stderr,
		    "%s: %s: only owner may touch special files on this filesystem\n",
		    prog, filename);
		return (-1);
	}

#ifndef S5EMUL
	if (nowrite && force) {
		if (chmod(filename, 0666)) {
			fprintf(stderr, "%s: could not chmod ", prog);
			perror(filename);
			return (-1);
		}
		rwstatus = readwrite(filename, statp->st_size);
		if (chmod(filename, (int)statp->st_mode)) {
			fprintf(stderr, "%s: could not chmod back ", prog);
			perror(filename);
			return (-1);
		}
		return (rwstatus);
	} else
#endif /*!S5EMUL*/
		return (readwrite(filename, statp->st_size));
}

int
readwrite(filename, size)
	char	*filename;
	off_t	size;
{
	int fd;
	char first;

	if (size) {
		if ((fd = open(filename, 2)) < 0)
			goto error;
		if (read(fd, &first, 1) != 1)
			goto closeerror;
		if (lseek(fd, 0L, 0) == -1)
			goto closeerror;
		if (write(fd, &first, 1) != 1)
			goto closeerror;
	} else {
		if ((fd = creat(filename, 0666)) < 0)
			goto error;
	}
	if (close(fd) < 0)
		goto error;
	return (0);

closeerror:
	(void) close(fd);
error:
	(void) fprintf(stderr, "%s: could not touch ", prog);
	perror(filename);
	return (-1);
}

#ifdef S5EMUL
isnumber(s)
	char *s;
{
	char c;

	while (c = *s++)
		if (!isdigit(c))
			return (0);
	return (1);
}
#endif /* S5EMUL */

/*
 * nanoseconds are rounded off to microseconds by flooring.
 */
static void
timestruc_to_timeval(timestruc_t *ts, struct timeval *tv)
{
	tv->tv_sec = ts->tv_sec;
	tv->tv_usec = ts->tv_nsec / 1000;
}
